js/misc/paygManager.js

```diff
@@ -190,6 +239,73 @@ var PaygManager = new Lang.Class({
         this.emit('code-expired');
     },
 
+    _timeRemainingSecs: function() {
+        if (!this._enabled)
+            return GLib.MAXUINT64;
+
+        return Math.max(0, this._expiryTime - (GLib.get_real_time() / GLib.USEC_PER_SEC));
+    },
+
+    _notifyPaygReminder: function(secondsLeft) {
+        // Only notify when in an regular session, not in GDM or initial-setup.
+        if (Main.sessionMode.currentMode != 'user' &&
+            Main.sessionMode.currentMode != 'user-coding') {
+            return;
+        }
+
+        if (this._notification)
+            this._notification.destroy();
+
+        let source = new MessageTray.SystemNotificationSource();
+        Main.messageTray.add(source);
+
+        let title = timeToString(secondsLeft);
+        this._notification = new MessageTray.Notification(source,
+                                                          NOTIFICATION_TITLE_FORMAT_STRING.format(title),
+                                                          NOTIFICATION_DETAILED_TEXT);
+        this._notification.setUrgency(MessageTray.Urgency.HIGH);
+        this._notification.setTransient(false);
+        source.notify(this._notification);
+
+        this._notification.connect('destroy', function() {
+            this._notification = null;
+        });
+    },
+
+    _updateExpirationReminders: function() {
+        if (this._expirationReminderId > 0) {
+            Mainloop.source_remove(this._expirationReminderId);
+            this._expirationReminderId = 0;
+        }
+
+        let secondsLeft = this._timeRemainingSecs();
+        if (secondsLeft <= 0)
+            return;
+
+        // Look for the right time to set the alarm for.
+        let targetAlertTime = 0;
+        for (let alertTime of notificationAlertTimesSecs) {
+            if (secondsLeft > alertTime) {
+                targetAlertTime = alertTime;
+                break;
+            }
+        }
+
+        // Too late to set up an alarm now.
+        if (targetAlertTime == 0)
+            return;
+
+        this._expirationReminderId = Mainloop.timeout_add_seconds(secondsLeft - targetAlertTime, () => {
+            // We want to show "round" numbers in the notification, matching
+            // whatever is specified in the notificationAlertTimeSecs array.
+            this._notifyPaygReminder(targetAlertTime);
+            this._updateExpirationReminders();
+
+            this._expirationReminderId = 0;
```

Comment:
Good catch. I didn't realize because in my mind I was setting the ID to `0` at the time I was removing the source (by returning `G_SOURCE_REMOVE`), but the fact that this is calling `_updateExpirationReminders()` again causes the problem you mention.

I'll change that.